diff --git a/dlls/mshtml/nsiface.idl b/dlls/mshtml/nsiface.idl
index 8c11032..0f09b80 100644
--- a/dlls/mshtml/nsiface.idl
+++ b/dlls/mshtml/nsiface.idl
@@ -3016,8 +3016,8 @@ interface nsIParser : nsISupports
     void Reset();
     PRBool CanInterrupt();
     PRBool IsInsertionPointDefined();
-    void BeginEvaluatingParserInsertedScript();
-    void EndEvaluatingParserInsertedScript();
+    void __cdecl BeginEvaluatingParserInsertedScript();
+    void __cdecl EndEvaluatingParserInsertedScript();
     void MarkAsNotScriptCreated();
     PRBool IsScriptCreated();
 }
