--- libs/wine/config.c.orig	2009-11-26 23:36:37.000000000 +0100
+++ libs/wine/config.c	2009-11-26 23:37:36.000000000 +0100
@@ -35,9 +35,9 @@
 #endif
 #include "wine/library.h"
 
-static const char server_config_dir[] = "/.wine";        /* config dir relative to $HOME */
-static const char server_root_prefix[] = "/tmp/.wine-";  /* prefix for server root dir */
-static const char server_dir_prefix[] = "/server-";      /* prefix for server dir */
+static const char server_config_dir[] = "/.wine-x86_64";        /* config dir relative to $HOME */
+static const char server_root_prefix[] = "/tmp/.wine-x86_64-";  /* prefix for server root dir */
+static const char server_dir_prefix[] = "/server-x86_64-";      /* prefix for server dir */
 
 static char *bindir;
 static char *dlldir;
