--- configure.orig	2006-08-28 06:44:28.000000000 +0200
+++ configure	2006-08-28 06:46:41.000000000 +0200
@@ -15903,7 +15903,7 @@
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_wine_try_cflags_saved=$CFLAGS
-CFLAGS="$CFLAGS -fPIC -Wl,--rpath,\$ORIGIN/../lib"
+CFLAGS="$CFLAGS -fPIC"
 cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
 _ACEOF
@@ -15969,7 +15969,7 @@
 echo "${ECHO_T}$ac_cv_ld_rpath" >&6; }
       if test "$ac_cv_ld_rpath" = "yes"
       then
-        LDEXERPATH="-Wl,--rpath,\\\$\$ORIGIN/\`\$(RELPATH) \$(bindir) \$(libdir)\`"
+        LDEXERPATH=""
 
       fi
 
--- configure.ac.orig	2006-08-28 06:43:35.000000000 +0200
+++ configure.ac	2006-08-28 06:44:22.000000000 +0200
@@ -1121,14 +1121,6 @@
         AC_SUBST(LDEXECFLAGS,["-Wl,--export-dynamic"])
       fi
 
-      AC_CACHE_CHECK([whether the linker accepts --rpath], ac_cv_ld_rpath,
-        [WINE_TRY_CFLAGS([-fPIC -Wl,--rpath,\$ORIGIN/../lib],
-                         ac_cv_ld_rpath="yes",ac_cv_ld_rpath="no")])
-      if test "$ac_cv_ld_rpath" = "yes"
-      then
-        AC_SUBST(LDEXERPATH,["-Wl,--rpath,\\\$\$ORIGIN/\`\$(RELPATH) \$(bindir) \$(libdir)\`"])
-      fi
-
       case $host_cpu in
         *i[[3456789]]86* | x86_64)
           AC_CACHE_CHECK([whether we can relocate the executable to 0x7bf00000], ac_cv_ld_reloc_exec,
