--- tools/wine.desktop.orig	2008-07-28 20:36:29.000000000 +0200
+++ tools/wine.desktop	2008-07-28 20:36:43.000000000 +0200
@@ -16,5 +16,5 @@
 Name[nb]=Wine Programlaster for Windowsapplikasjoner
 Name[nn]=Wine Programlaster for Windowsapplikasjoner
 Exec=wine start /unix %f
-MimeType=application/x-ms-dos-executable;application/x-msdos-program;application/x-msdownload;application/exe;application/x-exe;application/dos-exe;vms/exe;application/x-winexe;application/msdos-windows;application/x-zip-compressed;application/x-executable;application/x-msi;
+MimeType=application/x-ms-dos-executable;application/x-msdos-program;application/x-msdownload;application/exe;application/x-exe;application/dos-exe;vms/exe;application/x-winexe;application/msdos-windows;application/x-zip-compressed;application/x-msi;
 NoDisplay=true
