diff --git a/dlls/winealsa.drv/waveout.c b/dlls/winealsa.drv/waveout.c
index f4452e5..34ab57a 100644
--- a/dlls/winealsa.drv/waveout.c
+++ b/dlls/winealsa.drv/waveout.c
@@ -570,7 +570,7 @@ static DWORD wodOpen(WORD wDevID, LPWAVEOPENDESC lpDesc, DWORD dwFlags)
     snd_pcm_access_t            access;
     snd_pcm_format_t            format = -1;
     unsigned int                rate;
-    unsigned int                buffer_time = 500000;
+    unsigned int                buffer_time = 120000;
     unsigned int                period_time = 10000;
     snd_pcm_uframes_t           buffer_size;
     snd_pcm_uframes_t           period_size;

diff --git a/dlls/dsound/dsound_main.c b/dlls/dsound/dsound_main.c
index 31078cb..ad08fae 100644
--- a/dlls/dsound/dsound_main.c
+++ b/dlls/dsound/dsound_main.c
@@ -91,7 +91,7 @@ HRESULT mmErr(UINT err)
 /* All default settings, you most likely don't want to touch these, see wiki on UsefulRegistryKeys */
 int ds_emuldriver = 0;
 int ds_hel_buflen = 32768;
-int ds_snd_queue_max = 10;
+int ds_snd_queue_max = 12;
 int ds_snd_queue_min = 6;
 int ds_snd_shadow_maxsize = 2;
 int ds_hw_accel = DS_HW_ACCEL_FULL;
